dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/jdbc/JdbcCustomIconStore.java

```diff
@@ -0,0 +1,126 @@
+ * and/or other materials provided with the distribution.
+ * Neither the name of the HISP project nor the names of its contributors may
+ * be used to endorse or promote products derived from this software without
+ * specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.icon.jdbc;
+
+import java.util.List;
+
+import lombok.RequiredArgsConstructor;
+import lombok.extern.slf4j.Slf4j;
+
+import org.hisp.dhis.fileresource.FileResource;
+import org.hisp.dhis.icon.CustomIcon;
+import org.hisp.dhis.icon.CustomIconStore;
+import org.hisp.dhis.user.User;
+import org.springframework.jdbc.core.JdbcTemplate;
+import org.springframework.jdbc.core.RowMapper;
+import org.springframework.stereotype.Repository;
+
+@Slf4j
+@Repository( "org.hisp.dhis.icon.CustomIconStore" )
+@RequiredArgsConstructor
+public class JdbcCustomIconStore implements CustomIconStore
+{
+    private final JdbcTemplate jdbcTemplate;
+
+    @Override
+    public CustomIcon getIconByKey(String key) {
+        final String sql = """
+                    select c.key as iconkey, c.description as icondescription, c.keywords as keywords, f.uid as fileresourceuid, u.uid as useruid
+                    from customicon c join fileresource f on f.fileresourceid = c.fileresourceid
+                    join userinfo u on u.userinfoid = c.createdby
+                    where key = ?
+                    """;
+
+        List<CustomIcon> customIcons = jdbcTemplate.query(sql, customIconRowMapper, key);
+
+        return customIcons.isEmpty() ? null : customIcons.get(0);
+    }
+
+    @Override
+    public List<CustomIcon> getIconsByKeywords(String[] keywords) {
+        final String sql = """
+                    select c.key as iconkey, c.description as icondescription, c.keywords as keywords, f.uid as fileresourceuid, u.uid as useruid
+                    from customicon c join fileresource f on f.fileresourceid = c.fileresourceid
+                    join userinfo u on u.userinfoid = c.createdby
+                    where keywords @> string_to_array(?,',')
+                    """;
+
+        return jdbcTemplate.query(sql, customIconRowMapper, String.join(",", keywords));
```

Comment:
I had a comment here which I cannot see anymore so I don't know what your answer was 🤷🏻 It is awkward to join into a string and have to split again 😅 

have you looked at
https://www.baeldung.com/spring-jdbctemplate-in-list#NamedParameterJdbcTemplate

The driver does have support for `String[]` to postgres `text[]`
https://access.crunchydata.com/documentation/pgjdbc/42.2.20/arrays.html
which would make the joining and splitting a waste. Not sure if `JdbcTemplate` allows us to pass the `String[]` into `NamedParameterJdbcTemplate`. Prepared statements (used underneath in the JdbcTemplate AFAIK) seem to support arrays.